demo/bulk_insert/bulk_insert.py

```diff
@@ -1,281 +1,354 @@
+
+        self.clear_buffer()
+
+    # Delete all entities that have been inserted
+    def clear_buffer(self):
+        self.redis_token_count = 0
+        self.buffer_size = 0
+
+        # All constructed entities have been inserted, so clear buffers
+        self.node_count = 0
+        self.relation_count = 0
+        del self.labels[:]
+        del self.reltypes[:]
+
+# Convert a single CSV property field into a binary stream.
+# Supported property types are string, numeric, boolean, and NULL.
+def prop_to_binary(prop_str):
+    # All format strings start with an unsigned char to represent our Type enum
+    format_str = "=B"
+    if not prop_str:
+        # An empty field indicates a NULL property
+        return struct.pack(format_str, Type.NULL)
+
+    # If field can be cast to a float, allow it
+    try:
+        numeric_prop = float(prop_str)
+        return struct.pack(format_str + "d", Type.NUMERIC, numeric_prop)
+    except:
+        pass
+
+    # If field is 'false' or 'true', it is a boolean
+    if prop_str.lower() == 'false':
+        return struct.pack(format_str + '?', Type.BOOL, False)
+    elif prop_str.lower() == 'true':
+        return struct.pack(format_str + '?', Type.BOOL, True)
+
+    # If we've reached this point, the property is a string
+    # Encoding len+1 adds a null terminator to the string
+    format_str += "%ds" % (len(prop_str) + 1)
+    return struct.pack(format_str, Type.STRING, str.encode(prop_str))
+
+
+# Superclass for label and relation CSV files
+class EntityFile(object):
+    def __init__(self, filename):
+        # The label or relation type string is the basename of the file
+        self.entity_str = os.path.splitext(os.path.basename(filename))[0].encode("ascii")
+        # Input file handling
+        self.infile = open(filename, 'rt')
+        # Initialize CSV reader that ignores leading whitespace in each field
+        # and does not modify input quote characters
+        self.reader = csv.reader(self.infile, skipinitialspace=True, quoting=csv.QUOTE_NONE)
+
+        self.prop_offset = 0 # Starting index of properties in row
+        self.prop_count = 0 # Number of properties per entity
+
+        self.packed_header = ""
+        self.binary_entities = []
+        self.binary_size = 0 # size of binary token
+
+    # Simple input validations for each row of a CSV file
+    def validate_row(self, expected_col_count, row):
+        # Each row should have the same number of fields
+        if len(row) != expected_col_count:
+            raise CSVError("%s:%d Expected %d columns, encountered %d ('%s')"
+                           % (self.infile.name, self.reader.line_num, expected_col_count, len(row), ','.join(row)))
+        # Check for dangling commma
+        # TODO Since we do support empty CSV fields (interpreted as NULL values), this check may
+        # not be sensible anymore. It has the effect of allowing any property to be NULL except the last,
+        # which will cause this exception.
```

Comment:
@swilly22 What are your thoughts here? Dropping this check would standardize our property handling, so I think I lean towards that option. In practice, though, I think a trailing comma is usually indicative of a malformed CSV input.

Reply:
drop the check